test_util/test_launch.py

```diff
@@ -0,0 +1,302 @@
+
+import pytest
+import yaml
+from docopt import DocoptExit
+
+import ssh.tunnel
+import test_util
+from pkgpanda.util import load_json
+from test_util.helpers import Host
+from test_util.launch import get_launcher, LauncherError, main
+
+MOCK_SSH_KEY_DATA = 'ssh_key_data'
+MOCK_KEY_NAME = 'my_key_name'
+
+
+def check_cli(cmd):
+    assert main(cmd.split(' ')) == 0, 'CLI args failed! {}'.format(cmd)
+
+
+def check_success(capsys, tmpdir, config_str):
+    """
+    Runs through the required functions of a launcher and then
+    runs through the default usage of the script for a
+    given config path and info path, ensuring each step passes
+    if all steps finished successfully, this parses and returns the generated
+    info JSON and stdout description JSON for more specific checks
+    """
+    # Test launcher directly first
+    config = yaml.load(config_str)
+    launcher = get_launcher(config['type'], config['provider_info'])
+    info = launcher.create(config)
+    launcher.wait(info)
+    launcher.describe(info)
+    launcher.test(info, 'py.test')
+    launcher.delete(info)
+
+    # add config to disk and make info path for CLI testing
+    config_path = tmpdir.join('my_specific_config.yaml')  # test non-default name
+    config_path.write(config_str)
+    config_path = str(config_path)
+    info_path = str(tmpdir.join('my_specific_info.json'))  # test non-default name
+
+    # Now check launcher via CLI
+    check_cli('create --config-path={} --info-path={}'.format(config_path, info_path))
+    # use the info written to disk to ensure JSON parsable
+    info = load_json(info_path)
+    # General assertions about info
+    assert 'type' in info
+    assert 'provider' in info
+    assert 'ssh' in info
+    assert 'user' in info['ssh']
+    assert 'private_key' in info['ssh']
+
+    check_cli('wait --info-path={}'.format(info_path))
+
+    # clear stdout capture
+    capsys.readouterr()
+    check_cli('describe --info-path={}'.format(info_path))
+    # capture stdout from describe and ensure JSON parse-able
+    description = json.loads(capsys.readouterr()[0])
+
+    # general assertions about description
+    assert 'masters' in description
+    assert 'private_agents' in description
+    assert 'public_agents' in description
+
+    check_cli('test --info-path={}'.format(info_path))
+
+    check_cli('delete --info-path={}'.format(info_path))
+
+    return info, description
+
+
+@contextmanager
+def mock_context(*args, **kwargs):
+    yield type('Tunnelled', (object,), {})
+
+
+@pytest.fixture
+def mock_test_runner(monkeypatch):
+    monkeypatch.setattr(ssh.tunnel, 'tunnel', mock_context)
+    monkeypatch.setattr(test_util.runner, 'integration_test', lambda *args, **kwargs: 0)
+
+
+@pytest.fixture
+def mock_ssh_key_path(tmpdir):
+    ssh_key_path = tmpdir.join('ssh_key')
+    ssh_key_path.write(MOCK_SSH_KEY_DATA)
+    return str(ssh_key_path)
+
+
+@pytest.fixture
+def mock_aws_cf_simple_backend(monkeypatch, mock_test_runner):
```

Comment:
I cannot use pytest.mark.usefixtures in this case. Apparently this is only good for test cases because marking fixtures does not make a huge amount of sense in pytest. So don't try to tell anyone else to do this as it simply wont work :P